plenum/server/consensus/ordering_service.py

```diff
@@ -2166,6 +2171,25 @@ def _caught_up_till_3pc(self, last_caught_up_3PC):
         self.last_ordered_3pc = last_caught_up_3PC
         self._remove_till_caught_up_3pc(last_caught_up_3PC)
 
+        # Get all pre-prepares and prepares since the latest stable checkpoint
+        ledger = self.db_manager.get_ledger(AUDIT_LEDGER_ID)
+
+        to = ledger.get_last_txn()['txn']['data']['ppSeqNo']
+        frm = to - to % self._config.CHK_FREQ
+
+        batch_ids = [
+            BatchID(
+                view_no=self.view_no,  # TODO: Nemanja test passes if I put `txn['txn']['data']['viewNo']` instead of `self.view_no`. Why?
+                pp_view_no=txn['txn']['data']['viewNo'],
+                pp_seq_no=txn['txn']['data']['ppSeqNo'],
+                pp_digest=txn['txn']['data']['digest']
+            )
+            for _, txn in ledger.getAllTxn(frm=frm, to=to)
+        ]
+
+        self._data.preprepared = batch_ids[:]
```

Comment:
Why do we need to call `[:]`?

Reply:
If I don't use [:], preprepared and prepared will point to the same object, and an item added to one will show up in the other. So this is just a precaution, maybe an unnecessary one.